src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalescingAssignmentTests.cs

```diff
@@ -0,0 +1,1239 @@
+                //         i1 ??= 1.0;
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "i1 ??= 1.0").WithArguments("??=", "int?", "double").WithLocation(8, 9),
+                // (12,9): error CS0019: Operator '??=' cannot be applied to operands of type 'D' and 'C'
+                //         d1 ??= new C();
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "d1 ??= new C()").WithArguments("??=", "D", "C").WithLocation(12, 9),
+                // (16,9): error CS0019: Operator '??=' cannot be applied to operands of type 'C' and 'E'
+                //         c1 ??= new E();
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "c1 ??= new E()").WithArguments("??=", "C", "E").WithLocation(16, 9),
+                // (20,9): error CS0019: Operator '??=' cannot be applied to operands of type 'C' and 'F'
+                //         c2 ??= new F();
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "c2 ??= new F()").WithArguments("??=", "C", "F").WithLocation(20, 9)
+            });
+        }
+
+        [Fact]
+        public void AsInvalidSubexpression()
+        {
+            CreateCompilation(@"
+class C
+{
+    public void M()
+    {
+        double? d = 1.0;
+        M2(d ??= 2.0);
+        M3(d ??= 3.0);
+        object o = null;
+        M3(o ??= null);
+        (o ??= null) = null;
+    }
+
+    public void M2(double d) {}
+    public void M3(C c) {}
+}
+").VerifyDiagnostics(new DiagnosticDescription[] {
+                // (7,12): error CS1503: Argument 1: cannot convert from 'double?' to 'double'
+                //         M2(d ??= 2.0);
+                Diagnostic(ErrorCode.ERR_BadArgType, "d ??= 2.0").WithArguments("1", "double?", "double").WithLocation(7, 12),
+                // (8,12): error CS1503: Argument 1: cannot convert from 'double?' to 'C'
+                //         M3(d ??= 3.0);
+                Diagnostic(ErrorCode.ERR_BadArgType, "d ??= 3.0").WithArguments("1", "double?", "C").WithLocation(8, 12),
+                // (10,12): error CS1503: Argument 1: cannot convert from 'object' to 'C'
+                //         M3(o ??= null);
+                Diagnostic(ErrorCode.ERR_BadArgType, "o ??= null").WithArguments("1", "object", "C").WithLocation(10, 12),
+                // (11,10): error CS0131: The left-hand side of an assignment must be a variable, property or indexer
+                //         (o ??= null) = null;
+                Diagnostic(ErrorCode.ERR_AssgLvalueExpected, "o ??= null").WithLocation(11, 10)
+            });
+        }
+
+        [Fact]
+        public void DynamicLHS()
+        {
+            CompileAndVerify(@"
+using System;
+using System.Collections.Generic;
+public class C
+{
+    public static void Main()
+    {
+        VerifyField();
+        VerifyProperty();
+        VerifyIndexer();
+    }
+
+    public static void VerifyField()
+    {
+        Console.WriteLine(GetDynamicClass().F1 ??= 0);
```

Comment:
>Console.WriteLine [](start = 8, length = 17)

If it's necessary to verify the IL for each of these methods, consider moving `Console.WriteLine` to the caller in each case to reduce the size of the IL blocks.

Reply:
I considered this, but I wanted to verify using the assignment as an expression.

---
In reply to: [199839853](https://github.com/dotnet/roslyn/pull/28253#discussion_r199839853) [](ancestors = 199839853)

Reply:
Would `return GetDynamicClass().F1 ??= 0;` meet that requirement?

---
In reply to: [199930325](https://github.com/dotnet/roslyn/pull/28253#discussion_r199930325) [](ancestors = 199930325,199839853)

Reply:
It would, I'll update it to use that instead

---
In reply to: [199941799](https://github.com/dotnet/roslyn/pull/28253#discussion_r199941799) [](ancestors = 199941799,199930325,199839853)